Fetch some DBus properties through DBus method

jbaublitz opened the pull request:
This is a preliminary PR to resolve stratis-storage/project#52.

mulkieran commented on src/dbus_api/pool.rs:
```diff
@@ -485,41 +490,21 @@ pub fn create_dbus_pool<'a>(
-        .on_get(get_space_state);
+    let get_all_properties_method = f
+        .method("GetAllProperties", (), get_all_properties)
+        // b: true if filesystems were created
```
Copy/paste errors in comments and return type which I'm ignoring for now.

jbaublitz replied:
Resolved.

mulkieran commented on src/dbus_api/pool.rs:
```diff
@@ -360,48 +360,53 @@ fn get_pool_name(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result
-fn get_pool_state(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result<(), MethodErr> {
-    get_pool_property(i, p, |(_, _, pool)| Ok(pool.state() as u16))
-}
+    let (tpu_success, total_physical_used) = match pool.total_physical_used() {
```
We'll drop total physical used from the stats, as it is ill-defined.

jbaublitz replied:
Oops! Sorry, misunderstood the issue you assigned me. I'll remove this.

jbaublitz replied:
Resolved.

mulkieran commented on src/dbus_api/pool.rs:
```diff
@@ -360,48 +360,53 @@ fn get_pool_name(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result
+        consts::POOL_TOTAL_SIZE_PROP.to_string(),
+        (
+            true,
+            Variant(Box::new(format!("{}", pool.total_physical_size()))),
```
Use ```to_string()``` here.

jbaublitz replied:
Resolved.

mulkieran commented on src/dbus_api/pool.rs:
```diff
@@ -360,48 +360,53 @@ fn get_pool_name(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result
+        consts::POOL_TOTAL_USED_PROP.to_string(),
+        (tpu_success, Variant(Box::new(total_physical_used))),
+    );
+    return_value.insert(
```
I think we will want to drop some or all of these states as well as they relate to internal details that may be changed in future, rendering them meaningless.

jbaublitz commented:
The failure in stratisd-nonrust is due to the removal of properties in this PR. I'll fix those up.

jbaublitz commented:
Now the udev tests are failing. It's unclear to me whether or not this is related to my PR. My guess is that it is due to the failures just having started. However, I don't see anywhere that the udev tests depend on the properties that I removed.

jbaublitz commented:
This may actually not be due to this PR. See [here](https://github.com/stratis-storage/stratisd/pull/1664#issuecomment-543196897).

mulkieran commented:
Ok, one possibility for the table type is ```String -> (bool, (bool, var))```. If the outermost bool is false, then that means "unrecognized key", and everything else can be ignored. The code ought to easily supply a default value. If the outer bool is true, but the inner is false, that means value could not be obtained. In that case, since the type is variant, we can freely put whatever we want as the value. Currently, let us put a default value, for simplicity, but in future we could put an error string or some other useful information, and call that an enhancement. Finallly, if both bools are true, we've got the value and we return that.

I think this would make it fairly easy for a client to consume. In the CLI something like:

```
(key_exists, (value_exists, value)) = map[key]
if !key_exists:
     # raise some exception, we have a bug OR
     # do fallback, we have a too-early version of stratisd OR ...
if !value_exists:
     # do reasonable action. If printing out, perhaps substitute something like "<unkown>"
else:
    # use the actual value as appropriate
```

That seems workable to me, and better than anything else I have come up with so far.

jbaublitz commented:
I've been thinking about it too this morning, and this suggestion seems far more reasonable and expressive than anything I came up with. I'll go with that for now.

mulkieran commented on src/dbus_api/pool.rs:
```diff
@@ -360,48 +360,75 @@ fn get_pool_name(i: &mut IterAppend, p: &PropInfo<MTFn<TData>, TData>) -> Result
+
+    let uuid = get_data!(object_path; default_return; return_message).uuid;
+    let (_, pool) = get_pool!(engine; uuid; default_return; return_message);
+
```
Please turn the array of properties into a set before filling the map. My goal is just to prevent some frivolously obnoxious person from requesting the same expensive value to the limit of the size of the D-Bus array. If there's a way when filling the HashMap to do this as well, so that a request for a value can be guaranteed to be satisfied exactly once, that's great to.

jbaublitz replied:
I'm a little confused about the benefit of this. This still doesn't stop the user from supplying a bunch of random values which could still cause problems. Is the rationale that this would at least prevent the engine from running the code to fetch actual properties multiple times?

mulkieran replied:
Thinking further, I don't think it's necessary to return any entry for an unrecognized key. Let the fact that there is no entry for the key in the table be the indicator that the key was unrecognized.

Then the client does:
```
result = map.get(key)
if result is None:
   # raise some exception, we have a bug OR ...
(value_exists, value) = result
if !value_exists:
   # do reasonable action
else:
  # got the value, so use it
```

That makes the type of the table simpler, guarantees that it is no larger than the size of the set of keys that the daemon does recognize, and makes the type of the table the same as that for ```get_all_properties```.

Regarding your previous question, 2^20 keys will require only the computation necessary to realize that they are unrecognized, and in my new formulation won't lead to a very large table, which is also good. But, if the same recognized key is requested 2^20 times, I'ld like to avoid computing the value for that key 2^20 times, and just compute it once.

jbaublitz replied:
Resolved using `itertools`.

jbaublitz commented:
@mulkieran I've been poking around more and I have a prototype for reusing existing code across properties and methods to perform operations on a filesystem object. Take a look and see whether this seems reasonable. I made a few tweaks to the error type of the closure so that we can adequately capture the error message and return it as a `Variant` value in `get_all_properties` as opposed to bubbling up a `MethodErr`.

jbaublitz commented:
Fixing copy/paste error.

mulkieran commented:
> @mulkieran I've been poking around more and I have a prototype for reusing existing code across properties and methods to perform operations on a filesystem object. Take a look and see whether this seems reasonable. I made a few tweaks to the error type of the closure so that we can adequately capture the error message and return it as a `Variant` value in `get_all_properties` as opposed to bubbling up a `MethodErr`.

@jbaublitz We are trying to avoid having any ```StratisResult``` or ```StratisError``` generated in the D-Bus layer, we are trying to stick to only generating D-Bus errors there. What about making the return type of the ```filesystem_operation``` method just a ```Result```?

mulkieran commented on tests/client-dbus/tests/dbus/filesystem/test_properties.py:
```diff
@@ -78,9 +79,12 @@ def testProps(self):
 
-        self.assertEqual(used, "12345678")
+        self.assertEqual(used_success, True)
+        self.assertEqual(used, "12345678 bytes")
```
I don't like that the result contains "bytes", we should be sending the bare number over the D-Bus. To fix this, you have to dereference the value, so it isn't Bytes but just a number and _then_ stringify it. We should make sure to do that consistently everywhere.

jbaublitz replied:
In that case, I'd argue it should be an integer instead of a string. If we're not including units, I don't see a reason to encode it as a string.

mulkieran replied:
The reason it was a string was that the integers only go up to u64. With dedup and compression certain sizes that we report could be ridiculously large, larger than would fit in a 64 bit integer. So it is safer long-term to use the string.

jbaublitz replied:
Resolved.

jbaublitz commented:
@mulkieran Everything you've highlighted is resolved and I've pushed the `BlockDev` implementation for the properties. Like with pool, I've removed the `BlockDev` state property and moved size over to `FetchProperties`. This should be ready for final review.

jbaublitz commented:
Whoops! Resolved.

jbaublitz commented on src/dbus_api/event_handler.rs:
```diff
@@ -63,23 +63,6 @@ impl EngineListener for EventHandler {
                     });
                 }
             }
-            EngineEvent::PoolExtendStateChanged { dbus_path, state } => {
```
Resolved.

mulkieran commented on src/dbus_api/blockdev.rs:
```diff
@@ -177,52 +190,114 @@ fn set_user_info(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
+
+    let mut return_value: HashMap<String, (bool, Variant<Box<dyn RefArg>>)> = HashMap::new();
+
+    let bd_size_result = blockdev_operation(m.tree, object_path.get_name(), |_, bd| Ok(*bd.size()));
```
Ideally, we would abstract the shared computation between these two methods into a single computation. The only difference in behavior is that the properties are unknown in the second, but known in the first. So, maybe it would be possible to come up w/ something where the properties get set up in ```get_all_properties``` but just used in ```get_properties``` and the rest is shared.

Seems like: ```vec!["TotalPhysicalSize"].iter()``` would get you to almost the same state in ```get_all_properties``` as ```properties.unique()``` gets you in ```get_properties```, and I bet ```filter_map``` isn't terribly expensive.

jbaublitz replied:
Resolved.

mulkieran commented on src/dbus_api/blockdev.rs:
```diff
@@ -177,52 +190,114 @@ fn set_user_info(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
+    F: Fn(BlockDevTier, &dyn BlockDev) -> Result<R, String>,
+    R: dbus::arg::Append,
+{
+    let object_path = p.path.get_name();
```
Minor: I like to have as few identifiers as possible, so I avoid them if they're only used once; here you could remove the ```object_path``` identifier altogether and probably get nicer indentation.

NOTE: Applies to ```filesystem.rs``` and ```pool.rs``` as well.

jbaublitz replied:
Resolved.

mulkieran commented on src/dbus_api/pool.rs:
```diff
@@ -6,14 +6,15 @@ use std::{collections::HashMap, path::Path, vec::Vec};
     },
     Message,
 };
-
```
This looks like a stray line add, I hope its not rustfmt again.

jbaublitz replied:
This may be due to a nightly rustfmt, potentially the same as the `.clone()` change I made.

jbaublitz replied:
Should I change this back?

mulkieran replied:
Yes, please do.

jbaublitz replied:
Oh! I just looked at this more closely. The grouping of imports was inconsistent here so I was grouping all of the crate imports as one unit. I'll change it back if you don't like that.

mulkieran commented on src/engine/engine.rs:
```diff
@@ -16,8 +16,7 @@ use devicemapper::{Bytes, Device, Sectors};
         BlockDevState, BlockDevTier, CreateAction, DeleteAction, DevUuid, FilesystemUuid,
-        FreeSpaceState, MaybeDbusPath, Name, PoolExtendState, PoolState, PoolUuid, RenameAction,
-        SetCreateAction, SetDeleteAction,
+        MaybeDbusPath, Name, PoolUuid, RenameAction, SetCreateAction, SetDeleteAction,
```
You should be able to remove ```BlockDevState``` and ```BlockDev::state```, I think.

jbaublitz replied:
Resolved.

mulkieran commented on src/engine/strat_engine/tests/real.rs:
```diff
@@ -181,7 +181,7 @@ fn make_linear_test_dev(devnode: &Path, start: Sectors, length: Sectors) -> Line
         get_dm(),
         DmName::new(&format!("stratis_test_{}", Uuid::new_v4())).expect("valid format"),
         None,
-        table.clone(),
```
I think this is a stray edit, although it must be correct. Can you take it out of this PR and push it to a PR for November?

jbaublitz replied:
I added this due to accidentally using nightly for clippy and finding out we'll get an error soon. Do you still want me to separate this out?

mulkieran replied:
Go clippy! Yes please, toss it into November.

jbaublitz replied:
Reverted.

mulkieran commented on src/engine/strat_engine/thinpool/thinpool.rs:
```diff
@@ -621,23 +615,13 @@ impl ThinPool {
             new_state = PoolExtendState::MetaFailed;
         }
-        if self.extend_state() != new_state {
+        if self.pool_extend_state != new_state {
```
Please condense this to ```self.pool_extend_state = new_state``` as you did w/, e.g., ```free_space_state```.

mulkieran replied:
Exactly!

jbaublitz replied:
Resolved.

jbaublitz commented:
I found one lingering issue of `.append3()` instead of `.append1()` in the return value of `.get_all_properties()` for pool. I'm fixing that now.

jbaublitz commented:
Shared code has now been collapsed. I've avoiding collecting before passing it to `get_properties_shared` methods as we don't want to iterate through before calling `.unique()`. That feels like a waste.

mulkieran commented on src/dbus_api/pool.rs:
```diff
@@ -6,14 +6,15 @@ use std::{collections::HashMap, path::Path, vec::Vec};
 };
-
 use devicemapper::Sectors;
+use itertools::Itertools;
```
We have a special ordering for imports, first from standard library, then external crates furnished by others, then external crates furnished by ourselves, then local imports. So itertools import should be above devicemapper import.
